contracts/exchange/mixins/DcvSale.sol

```diff
@@ -0,0 +1,243 @@
+// SPDX-License-Identifier: Unlicense
+pragma solidity 0.8.19;
+
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+import "./DcvExchangeCore.sol";
+import "./DcvFundsDistributor.sol";
+import "./DcvEIP712.sol";
+import "./DcvNonceManager.sol";
+
+abstract contract DcvSale is
+    ReentrancyGuardUpgradeable,
+    DcvExchangeCore,
+    DcvFundsDistributor,
+    DcvEIP712,
+    DcvNonceManager
+{
+    using ECDSA for bytes32;
+
+    bytes32 private constant SALE_TYPEHASH =
+        keccak256(
+            "Sale(uint256 orderNonce,address nftContract,uint256 tokenId,uint256 price)"
+        );
+    bytes32 private constant SALE_WITH_ROYALTY_TYPEHASH =
+        keccak256(
+            "SaleWithRoyalty(uint256 orderNonce,address nftContract,uint256 tokenId,uint256 price,address royaltyRecipient,uint256 royaltyPercentage)"
+        );
+    bytes32 private constant ROYALTY_PARAMETERS_TYPEHASH =
+        keccak256(
+            "RoyaltyParameters(address nftContract,uint256 tokenId,address royaltyRecipient,uint256 royaltyPercentage)"
+        );
+
+    struct Sale {
+        address seller;
+        uint256 orderNonce;
+        address nftContract;
+        uint256 tokenId;
+        uint256 price;
+    }
+
+    struct SaleWithRoyalty {
+        address seller;
+        uint256 orderNonce;
+        address nftContract;
+        uint256 tokenId;
+        uint256 price;
+        address royaltyRecipient;
+        uint256 royaltyPercentage;
+    }
+
+    event SaleSuccess(
+        address indexed nftContract,
+        uint256 indexed tokenId,
+        address indexed seller,
+        address buyer,
+        uint256 price,
+        uint256 platformFee
+    );
+
+    function buyFromSale(
+        bytes calldata signature,
+        Sale calldata sale
+    ) external payable nonReentrant {
```

Comment:
Kako si zamislio buyFromSale i buyFromSaleWithRoyalties? Zasto imamo 2 metode?

Reply:
Posto imamo prodaju u kojoj cemo postavljat royalties parametre za taj token (kad prodaje originalni minter), to bi ovdje bila funkcija `buyFromSaleWithRoyalties`. Te imamo prodaju u kojoj cemo samo procitat vec postavljeni royalties za taj token (kad ne prodaje originalni minter), to bi ovdje bila funkcija `buyFromSale`. Mozda se moglo objedinit u jednu funkciju, al odvojia sam u dvi funkcije jer vjerojatno ce bit vise prodaja od neoriginalnog mintera, pa da se ne moraju uvik specificirati royalties parametri koji se ionako nece postaviti.